### ISSUE TITLE
checklist for improving console accessibility

### ISSUE THREAD

**Issue by jwforres** [openshift/origin#5988]
### Overview / Pod templates / Status and Scaling Donuts
- [x] scale down link has no tooltip when disabled, we should also properly disable the link, right now its just visually disabled
- [x] pod status has nothing except "Pod status:" in the screen reader only content, suspect something broke here ( fixed in https://github.com/openshift/origin/pull/6117 )
- [ ] overview view switcher is not keyboard accessible
- [ ] topology view not intended for screen reader usage, we need some sr-only content that says that, we probably have equivalent facilitation with `oc status` or with the structure we give you on the main overview
- [x] </> source icon in pod template probably needs title text or an aria-describe attribute since we dont say "source" anywhere (fixed in https://github.com/openshift/origin/pull/6417)
- [x] keyboard only users don't have a link to go to the pods for a deployment, should we make the whole donut svg area be wrapped in an <a>?
- [ ] Need to see how our project dropdown behaves with a screen reader
- [ ] Need to see how our label filter typeahead behaves with a screen reader
#### Catalog
- [x] tags on the tiles should have titles that say Filter by tag <tag_name> ( fixed in https://github.com/openshift/origin/pull/6120 )
- [x] the builder icon is aria-hidden, we should have an sr-only element that has the same help text or something... referring to the "This image can build source" text ( fixed in https://github.com/openshift/origin/pull/6120 )
#### From source + builder
- [ ] help icons on the forms next to the sections either need alt text or aria-hidden, aria-hidden is probably fine in this case since the link next to it says "About <whatever>"
- [ ] env var and label inputs may need either title or aria-describe attrs on the inputs since they don't have labels, unless the screen reader reads the placeholder text
- [ ] edit and delete buttons on the env var and label forms probably need to say "Edit <key>" and "Delete <key>" so if the env var's key was foo the buttons would be Edit foo and Delete foo
- [ ] edit value input field for env vars / labels needs title text or aris-describe
- [ ] replicas label needs the for attr and shouldn't be nesting the input underneath it
#### From template
- [x] parameters, the description should have an aria-describes set to the id of the input or something similar, so that the addtl description is read ( fixed in https://github.com/openshift/origin/pull/6123 )
- [ ] parameters, check if placeholder text is read, if not add it to an aria-describe 
### Browse
- [ ] links to the various resource pages within the tables probably need more descriptive title text for someone just tabbing through
- [ ] now that logs use table markup for layout we should add the aria indicator that it is not a data table
- [ ] the follow log and go to top links should possibly be aria-hidden since they just visually scroll the page content
- [x] our tabs are not in the keyboard navigation because they are missing an href, is this an issue with <tabset> ? (fixed in https://github.com/openshift/origin/pull/6417)
- [ ] annotation table needs a hidden header with column headers for "Annotation key" and "Annotation value"
- [ ] labels needs an sr-only text before it that says "Labels:" since we made them stand out visually otherwise
### General
- [ ] we should add alternate aria descriptions anywhere we use unicode arrows (in pod template, services, next to Try it link on create)
- [ ] we need to sweep everywhere we are using the warning icon, and our other alerts, to make sure we have alt text on the icons.  there are definitely places missing it.
- [ ] when we use the external link indicator, should we add some alternative text that tells them its an external link
- [x] everywhere we've got delete icons (trashcan) its an <i> and not in the keyboard nav, either needs tabindex or we need to wrap it in an <a> ( fixed in https://github.com/openshift/origin/pull/6152 )
- [ ] we should add a sr-only Skip main navigation link
### Metrics

??? Need to evaluate the metrics tab in more detail
- [ ] guessing the sparkline charts are not super accessible, probably need equivalent facilitation of some kind? sr-only table with the same data?
### Create Project
- [ ] check if placeholder text is being read
- [ ] name field has an additional description, need to associate it with the input
### Edit Build Config

Need to evaluate
- [ ] labels missing `for` attribute

**Comment by jwforres** [comment]
cc @benjaminapetersen @spadgett @liggitt @fabianofranz @jhadvig @sg00dwin dumped this in one large checklist for now, we should look at breaking off smaller chunks of this when we are doing other work in these areas.

**Comment by spadgett** [comment]
> scale down link has no tooltip when disabled, we should also properly disable the link, right now its just visually disabled

Should be able to "disable" the link like this:

```
<a href="" role="button" aria-disabled="true">...</a>
```

**Comment by spadgett** [comment]
Should we use `ng-aria`?

https://docs.angularjs.org/api/ngAria

**Comment by benjaminapetersen** [comment]
Looks like a good idea to me!

**Comment by jwforres** [comment]
On Mon, Nov 30, 2015 at 1:48 PM, Sam Padgett <notifications@github.com>
wrote:

> Should we use ng-aria?
>

Yes please.

**Comment by spadgett** [comment]
> everywhere we've got delete icons (trashcan) its an <i> and not in the keyboard nav, either needs tabindex or we need to wrap it in an <a>

Fixed by #6152

**Comment by spadgett** [comment]
Some small accessibility fixes to alerts in #6473.

**Comment by spadgett** [comment]
Related to #6849

**Comment by spadgett** [comment]
@jwforres added a checklist item for editing build configs

**Comment by jwforres** [comment]
As part of any accessibility work we should enable an accessibility linter to validate the templates are at least valid.  See the old issue https://github.com/openshift/origin/issues/6849

**Comment by benjaminapetersen** [comment]
Inclined to think we could migrate the remaining open items to openshift-web-console on this one as well.  I'm happy to do so if nobody objects.

**Comment by jwforres** [comment]
sure, check the comments for any other additions/deletions to the list

**Comment by jwforres** [comment]
I'm moving this to trello instead, this is a story that will need to be planned / prioritized https://trello.com/c/CeYg80aG
